lib/msf/base/sessions/command_shell.rb

```diff
@@ -312,18 +312,20 @@ def cmd_shell(*args)
   end
 
   #
-  # Check if there is a binary in PATH env
+  # Returns path of a binary in PATH env.
   #
   def binary_exists(binary)
     print_status("Trying to find binary(#{binary}) on target machine")
-    binary_path = shell_command_token("which #{binary}").to_s.strip
-    if binary_path.eql?("#{binary} not found") || binary_path == ""
-      print_error(binary_path)
-      return nil
-    else
-      print_status("Found #{binary} at #{binary_path}")
-      return binary_path
+    binary_path = shell_command_token("command -v '#{binary}'").to_s.strip
+    if binary_path.eql?("")
+      binary_path = shell_command_token("which #{binary}").to_s.strip
+      if binary_path.eql?("#{binary} not found") || binary_path == ""
+        print_error("#{binary} not found")
+        return nil
+      end
```

Comment:
The output of shells when they don't find a command can be variable (which is why this method exists), although this didn't seem to be handling `which` output well, either. A better set of tests would look like:
```suggestion
    if shell_command_token('command -v command').to_s.strip == 'command'
      binary_path = shell_command_token("command -v '#{binary}'").to_s.strip
    else
      binary_path = shell_command_token("which #{binary}").to_s.strip
    end
    if binary_path.end_with?("#{binary} not found") || binary_path.include?("no #{binary} in") || binary_path == ""
      print_error("#{binary} not found")
      return nil
    end
```

Reply:
Okay 👍

Reply:
Some checks failing will fix them.

Reply:
Hey @acammack-r7 :-)
Error messages like "not found" might vary with locales, which was mentioned in the original issue also.
```ruby
if binary_path.end_with?("#{binary} not found") || binary_path.include?("no #{binary} in") || binary_path == ""
```
I've made some changes to your suggestions. Please have a look at it.
```ruby
if shell_command_token('command -v command').to_s.strip == 'command'
  binary_path = shell_command_token("command -v '#{binary} && true'").to_s.strip
else
  binary_path = shell_command_token("which '#{binary}' && echo true").to_s.strip
end
unless binary_path.include?("true")
  print_error("#{binary} not found")
  return nil
end
binary_path = binary_path.split("\n")[0].strip  #removes 'true' from stdout
print_status("Found #{binary} at #{binary_path}")
return binary_path
```

Reply:
not able to figure out why it's failing...